### PR TITLE
CORE-6082 named query integration into flows

### PR DESCRIPTION
**Pull request by LWogan**
https://github.com/corda/corda-api/pull/527

### CODE REVIEW

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/NamedParameterisedQuery.kt [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import java.nio.ByteBuffer
+import net.corda.flow.application.persistence.external.events.NamedQueryExternalEventFactory
+import net.corda.flow.application.persistence.external.events.NamedQueryParameters
+import net.corda.flow.application.persistence.wrapWithPersistenceException
+import net.corda.flow.external.events.executor.ExternalEventExecutor
+import net.corda.flow.fiber.FlowFiberSerializationService
+import net.corda.v5.application.persistence.ParameterisedQuery
+
+/**
+ * Object used to set and execute named queries against a database
```
Comment (simon-johnson-r3) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/NamedParameterisedQuery.kt]: Pretty pedantic but this isn't an object.
Perhaps: "Instances of this class are..." ?
(also applies to other classes)

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/PagedQueryFactory.kt [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+class PagedQueryFactory @Activate constructor(
+    @Reference(service = ExternalEventExecutor::class)
+    private val externalEventExecutor: ExternalEventExecutor,
+    @Reference(service = FlowFiberSerializationService::class)
+    private val flowFiberSerializationService: FlowFiberSerializationService,
+) {
+
+    /**
+     * Create a parameterised query object to execute named queries against the external DB process.
+     * Sets default values of offset to 0, and limit to max int max int value. Sets query params to empty.
```
Comment (simon-johnson-r3) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/PagedQueryFactory.kt]: typo?: max int max int value

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/PagedQueryFactory.kt [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+     * @param queryName the name of the query to execute
+     * @return NamedParameterisedQuery object that can be used to execute named queries.
```
Comment (simon-johnson-r3) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/PagedQueryFactory.kt]: @param expectedClass missing

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/fiber/FlowFiberSerializationService.kt [hunk 4]
```diff
@@ -0,0 +1,42 @@
+package net.corda.flow.fiber
+
+import net.corda.v5.application.serialization.SerializationService
+import net.corda.v5.base.exceptions.CordaRuntimeException
+import net.corda.v5.serialization.SerializedBytes
+
+/**
+ * [FlowFiberSerializationService]
```
Comment (simon-johnson-r3) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/fiber/FlowFiberSerializationService.kt]: As far as I know this just creates a link the class without any further value? Is that what you want here? I would suggest nothing at all is better.

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/fiber/FlowFiberSerializationService.kt [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+     * @throws DeserializedWrongAMQPObjectException If [bytes] deserializes into an object that is not an
+     * [expectedType].
+     */
+    fun <R : Any> deserialize(bytes: ByteArray, expectedType: Class<R>): R
+
+    /**
+     * Serialize an object using the flow fiber's AMQP [SerializationService].
+     *
+     * @param obj The object to serialize.
+     *
```
Comment (simon-johnson-r3) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/fiber/FlowFiberSerializationService.kt:45]: inconsistent blank line

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/fiber/FlowFiberSerializationServiceImpl.kt [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        get(): SerializationService {
+            return flowFiberService.getExecutingFiber().getExecutionContext().run {
+                sandboxGroupContext.amqpSerializer
+            }
+        }
+
+    override fun <R : Any> deserialize(bytes: ByteArray, expectedType: Class<R>): R {
+        return try {
+            val payload = serializationService.deserialize(bytes, expectedType)
+            checkDeserializedObjectIs(payload, expectedType)
```
Comment (simon-johnson-r3) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/fiber/FlowFiberSerializationServiceImpl.kt]: checkNotNull (for example) returns the object if it passes the test, if you did that here you'd not need line 34 which looks a bit out on its own. Personal preference though, feel free to ignore.

File: components/flow/flow-service/src/test/kotlin/net/corda/flow/application/persistence/query/PagedQueryFactoryTest.kt [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    private class TestObject
+
+    private val externalEventExecutor = mock<ExternalEventExecutor>()
+    private val flowFiberSerializationService = mock<FlowFiberSerializationService>()
+    private val pagedQueryFactory = PagedQueryFactory(externalEventExecutor, flowFiberSerializationService)
+
+    @Test
+    fun `create named query`() {
+        assertThat(pagedQueryFactory.createNamedParameterisedQuery("query", TestObject::class.java)::class.java)
+            .isEqualTo(NamedParameterisedQuery::class.java)
```
Comment (simon-johnson-r3) [inline components/flow/flow-service/src/test/kotlin/net/corda/flow/application/persistence/query/PagedQueryFactoryTest.kt]: The function declaration of `createNamedParameterisedQuery` specifies it returns a NamedParameterisedQuery, so I'm not sure this test adds value?
You could probably rewrite it so the test wouldn't compile if this wasn't true.

File: components/flow/flow-service/src/test/kotlin/net/corda/flow/application/persistence/query/PagedQueryFactoryTest.kt [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    @Test
+    fun `create find paged query`() {
+        assertThat(pagedQueryFactory.createPagedFindQuery(TestObject::class.java)::class.java)
```
Comment (simon-johnson-r3) [inline components/flow/flow-service/src/test/kotlin/net/corda/flow/application/persistence/query/PagedQueryFactoryTest.kt:50]: Same here.

**Review by simon-johnson-r3 — DISMISSED** [review]

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/NamedParameterizedQuery.kt [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    private val externalEventExecutor: ExternalEventExecutor,
+    private val flowFiberSerializationService: FlowFiberSerializationService,
+    private val queryName: String,
+    private var parameters: MutableMap<String, Any>,
+    private var limit: Int,
+    private var offset: Int,
+    private var expectedClass: Class<R>,
+) : ParameterizedQuery<R> {
+
+    @Suspendable
```
Comment (owenstanford) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/NamedParameterizedQuery.kt]: Do these set methods need to be ```@Suspendable``` the code can't suspend inside them?

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/NamedParameterizedQuery.kt [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+            externalEventExecutor.execute(
+                NamedQueryExternalEventFactory::class.java,
+                NamedQueryParameters(queryName, getSerializedParameters(parameters), offset, limit)
+            )
+        }.map { flowFiberSerializationService.deserialize(it.array(), expectedClass) }
+
+        return deserialized
+    }
+
+    @Suspendable
```
Comment (owenstanford) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/NamedParameterizedQuery.kt]: Same here, is ```@Suspendable``` required

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/NamedParameterizedQuery.kt [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    override fun setParameters(parameters: Map<String, Any>): ParameterizedQuery<R> {
+        this.parameters = parameters.toMutableMap()
+        return this
+    }
+
+    @Suspendable
+    override fun execute(): List<R> {
+        val deserialized = wrapWithPersistenceException {
```
Comment (owenstanford) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/NamedParameterizedQuery.kt]: We're missing a test to assert the exception wrapping

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/PagedFindQuery.kt [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+ */
+class PagedFindQuery<R : Any>(
+    private val externalEventExecutor: ExternalEventExecutor,
+    private val flowFiberSerializationService: FlowFiberSerializationService,
+    private val entityClass: Class<R>,
+    private var limit: Int,
+    private var offset: Int,
+) : PagedQuery<R> {
+
+    @Suspendable
```
Comment (owenstanford) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/PagedFindQuery.kt]: ```@Suspendable```  on methods that don't suspend?

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/PagedFindQuery.kt [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+    override fun setOffset(offset: Int): PagedQuery<R> {
+        require (offset >= 0) { "Offset cannot be negative" }
+        this.offset = offset
+        return this
+    }
+
+    @Suspendable
+    override fun execute(): List<R> {
+        val deserialized = wrapWithPersistenceException {
```
Comment (owenstanford) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/PagedFindQuery.kt:36]: No test to validate exception wrapping

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/NamedParameterizedQuery.kt (hunk 9 above)
Comment (lankydan) [reply components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/NamedParameterizedQuery.kt]: While I agree, we've mainly decided to just put `@Suspendable` onto everything.

Comment (owenstanford) [reply components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/NamedParameterizedQuery.kt]: For methods that are calling other methods I can see why we might take that approach, but for simple setters and util methods that don't call anything else I'm not sure it makes sense. There is a danger that if we use them everywhere the purpose of them becomes unclear.

Comment (lankydan) [reply components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/NamedParameterizedQuery.kt]: True, they're not on the `PageQuery` interface anyway as it turns out, so I'll drop them.

**Comment by dickon** [comment]
I objected on the matching Corda-api pull request to the partial renaming to American spellings, so that same comment applies here to all the changes that follow from the rename.

**Review by owenstanford — DISMISSED** [review]

**Comment by lankydan** [comment]
@dickon I've removed the `ized` change from the API and runtime PRs. Will deal with it separately as its a more general thing.

**Review by driessamyn — DISMISSED** [review]
LGTM

**Review by driessamyn — APPROVED** [review]
LGTM

**Pull request merged**